From 83e46d0bdff0176dc4ee3cef75ff8df5a04b9b78 Mon Sep 17 00:00:00 2001
From: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Date: Tue, 29 Apr 2014 14:13:06 +0100
Subject: [PATCH 208/351] ASoC: wm_adsp: Set compressed capture watermark

Sets the firmware buffer watermark to match the
compressed stream fragment size.

Change-Id: I653989b9a0db49b629abc6439d6d3c1ca5e89458
Signed-off-by: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 sound/soc/codecs/wm_adsp.c |   20 ++++++++++++++++++++
 sound/soc/codecs/wm_adsp.h |    1 +
 2 files changed, 21 insertions(+)

diff --git a/sound/soc/codecs/wm_adsp.c b/sound/soc/codecs/wm_adsp.c
index 03c480a..2ab662b 100644
--- a/sound/soc/codecs/wm_adsp.c
+++ b/sound/soc/codecs/wm_adsp.c
@@ -286,6 +286,8 @@ struct wm_adsp_host_buffer {
 #define WM_ADSP_MAX_READ_SIZE          256
 #define WM_ADSP_ALG_XM_STRUCT_MAGIC    0x49aec7
 
+#define WM_ADSP_DEFAULT_WATERMARK      DIV_ROUND_UP(2048, WM_ADSP_DATA_WORD_SIZE)
+
 #define ADSP2_SYSTEM_CONFIG_XM_PTR \
 	(offsetof(struct wmfw_adsp2_id_hdr, xm) / sizeof(__be32))
 
@@ -2752,6 +2754,16 @@ int wm_adsp_stream_alloc(struct wm_adsp *adsp,
 		}
 	}
 
+	size = params->buffer.fragment_size;
+	if (size == 0) {
+		adsp->capt_watermark = WM_ADSP_DEFAULT_WATERMARK;
+		adsp_warn(adsp, "No fragment size, assuming %u",
+				adsp->capt_watermark * WM_ADSP_DATA_WORD_SIZE);
+	} else {
+		adsp->capt_watermark =
+				DIV_ROUND_UP(size, WM_ADSP_DATA_WORD_SIZE);
+	}
+
 	return 0;
 
 err_raw_capt_buf:
@@ -2826,6 +2838,14 @@ int wm_adsp_stream_start(struct wm_adsp *adsp)
 	if (ret < 0)
 		return ret;
 
+	ret = wm_adsp_host_buffer_write(adsp,
+					HOST_BUFFER_FIELD(high_water_mark),
+					adsp->capt_watermark);
+	if (ret < 0)
+		return ret;
+
+	adsp_dbg(adsp, "Set watermark to %u\n", adsp->capt_watermark);
+
 	return 0;
 }
 EXPORT_SYMBOL_GPL(wm_adsp_stream_start);
diff --git a/sound/soc/codecs/wm_adsp.h b/sound/soc/codecs/wm_adsp.h
index 2f07294..4212e88 100644
--- a/sound/soc/codecs/wm_adsp.h
+++ b/sound/soc/codecs/wm_adsp.h
@@ -100,6 +100,7 @@ struct wm_adsp {
 	u32 *raw_capt_buf;
 	struct circ_buf capt_buf;
 	int capt_buf_size;
+	u32 capt_watermark;
 	struct wm_adsp_buffer_region *host_regions;
 	bool buffer_drain_pending;
 
-- 
1.7.9.5

